src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,473 @@
+            Environment.GetEnvironmentVariable("ANDROID_ROOT") + "/usr/share/zoneinfo/tzdata",
+        };
+
+        private string tzdataPath;
+        private Stream? data;
+
+        private string[]? ids;
+        private int[]? byteOffsets;
+        private int[]? lengths;
+
+        public AndroidTzData(params string[] paths)
+        {
+            foreach (var path in paths)
+            {
+                if (LoadData(path))
+                {
+                    tzdataPath = path;
+                    return;
+                }
+            }
+
+            tzdataPath = "/";
+            ids = new[]{ "GMT" };
+        }
+
+        private static string GetApexTimeDataRoot()
+        {
+            string? ret = Environment.GetEnvironmentVariable("ANDROID_TZDATA_ROOT");
+            if (!string.IsNullOrEmpty(ret!)) {
+                return ret!;
+            }
+
+            return "/apex/com.android.tzdata";
+        }
+
+        private static string GetApexRuntimeRoot()
+        {
+            string? ret = Environment.GetEnvironmentVariable("ANDROID_RUNTIME_ROOT");
+            if (!string.IsNullOrEmpty(ret!))
+            {
+                return ret!;
+            }
+
+            return "/apex/com.android.runtime";
+        }
+
+        private bool LoadData(string path)
+        {
+            if (!File.Exists(path))
+                return false;
+
+            try
+            {
+                data = File.OpenRead(path);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                return false;
+            }
+
+            try
+            {
+                ReadHeader();
+                return true;
+            }
+            catch
+            {
+                // log something here instead of the console.
+                //Console.Error.WriteLine ("tzdata file \"{0}\" was present but invalid: {1}", path, e);
+            }
+            return false;
+        }
+
+        private unsafe void ReadHeader()
+        {
+            int size   = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+            var buffer = new byte[size];
+            var header = ReadAt<AndroidTzDataHeader>(0, buffer);
+
+            header.indexOffset = NetworkToHostOrder(header.indexOffset);
+            header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+            sbyte* s = (sbyte*)header.signature;
+            string magic = new string(s, 0, 6, Encoding.ASCII);
+
+            if (magic != "tzdata" || header.signature[11] != 0)
+            {
+                var b = new StringBuilder();
+                b.Append("bad tzdata magic:");
+                for (int i = 0; i < 12; ++i) {
+                    b.Append(' ').Append(((byte)s[i]).ToString("x2"));
+                }
```

Comment:
I'm not sure this extra code is worth it and it's wrong for signature[11] != 0

Reply:
I wasn't sure exactly how tz data headers should be read, so didn't yet modify the logic from the mono/mono implementation.

Reply:
/cc @grendello

Reply:
I think this code is to make `byte signature[12]` a string for exception throwing purposes. Why is it wrong for `signature[11] != 0`?